downstairs/src/lib.rs

```diff
@@ -1068,17 +1068,18 @@ impl ActiveConnection {
 
     /// Handle a new message from the upstairs
     ///
-    /// If the message is an IO, then put the new IO the work hashmap. If the
-    /// message is a repair message, then we handle it right here.
+    /// Returns `true` if the message was evaluated, `false` otherwise, or an
+    /// error if communication failed.
```

Comment:
How about returning an enum instead? Something like:

```rust
enum ProcFrameResult {
  WorkDone,
  WorkQueued,
}
```

It's less convenient because it can't be put right in an if conditional but it's less confusing at the call site.

Reply:
Good call, I added a new `WorkResult` type with those two variants (since it's used in `do_work_if_ready`, not just `proc_frame`)